### PR TITLE
Add initial OOM test infrastructure

### PR DESCRIPTION
**Pull request by fitzgen**
Baby steps towards https://github.com/bytecodealliance/wasmtime/issues/12069

### CODE REVIEW

File: crates/fuzzing/src/oom.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+            match old_state {
+                OomState::OutsideOomTest => {
+                    new_state = OomState::OutsideOomTest;
+                    ptr = unsafe { std::alloc::System.alloc(layout) };
+                }
+                OomState::OomOnAlloc(0) => {
+                    log::trace!(
+                        "injecting OOM for allocation: {layout:?}\nAllocation backtrace:\n{:?}",
+                        Backtrace::new(),
+                    );
```
Comment (alexcrichton) [inline crates/fuzzing/src/oom.rs:92]: Given how low-level allocation is and how high-level printing/backtraces are, I think it'd probably be best to remove this. I'd imagine that any error-related to OOM would already be aborting and/or something in a context where we could get a stack trace, so logging/printing a new one here seems ok to defer to that style of debugging. Basically I'd be worried that the `OutsideOomTest` is not sufficient to block any possible malloc anywhere to be able to print/backtrace on that call.

Could also wait to see if it's a problem too, and it may not end up being a problem.

File: crates/fuzzing/src/oom.rs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+                    new_state = OomState::DidOom;
+                    ptr = ptr::null_mut();
+                }
+                OomState::OomOnAlloc(c) => {
+                    new_state = OomState::OomOnAlloc(c - 1);
+                    ptr = unsafe { std::alloc::System.alloc(layout) };
+                }
+                OomState::DidOom => {
+                    panic!("OOM test attempted to allocate after OOM: {layout:?}")
+                }
```
Comment (alexcrichton) [inline crates/fuzzing/src/oom.rs:102]: This feels onerous to me to require which is to say that after one allocation fails all other allocations must be avoided entirely. I could imagine it being pretty reasonable that allocations are still attempted on some paths which are then also handled if they fail.

Would it make sense to hard-return null while `DidOom` is true to "permanently OOM"? Or maybe only force one OOM but let other allocations keep going?

File: crates/fuzzing/src/oom.rs [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+            match (result, old_state) {
+                (_, OomState::OutsideOomTest) => unreachable!(),
+
+                // The test function completed successfully before we ran out of
+                // allocation fuel, so we're done.
+                (Ok(()), OomState::OomOnAlloc(_)) => break,
+
+                // We injected an OOM and the test function handled it
+                // correctly; continue to the next iteration.
+                (Err(e), OomState::DidOom) if self.is_oom_error(&e) => {}
```
Comment (alexcrichton) [inline crates/fuzzing/src/oom.rs:215]: This feels perhaps a bit over-ambitious to me where we might have various edge cases where allocation failure can be handled. For example we might have a test saying "invoke a function twice" which invokes it a second time even if the first time failed, ensuring that an OOM on one call doesn't prevent some other call from happening (and reading corrupted state/etc). In such a situation the test would still return success, despite handling the OOM.

More generally given that the main goal here is "didn't abort on OOM" I'd think it'd be reasonable to transform this to "well the test returned to that means success, yay!"

File: crates/fuzzing/src/oom.rs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+            {
+                break;
+            }
+
+            log::trace!("=== Injecting OOM after {i} allocations ===");
+            let old_state = set_oom_state(OomState::OomOnAlloc(i));
+            assert_eq!(old_state, OomState::OutsideOomTest);
+
+            let result = test_func();
+            let old_state = set_oom_state(OomState::OutsideOomTest);
```
Comment (alexcrichton) [inline crates/fuzzing/src/oom.rs]: This'll likely want to be in a `Drop` to avoid heavily crashing the process by accident in the case that `test_func` panics

File: crates/fuzzing/src/oom.rs (hunk 1 above)
Comment (bjorn3) [reply crates/fuzzing/src/oom.rs:92]: Currently oom failure aborts without backtrace, just a panic message containing the amount of bytes it attempted to allocate.

Comment (alexcrichton) [reply crates/fuzzing/src/oom.rs:92]: Oh sure yeah, but that's a pretty easy event to break on in a debugger to get a backtrace

Comment (fitzgen) [reply crates/fuzzing/src/oom.rs:92]: > Could also wait to see if it's a problem too, and it may not end up being a problem.

I'd prefer to do this, just because, while it is always possible to attach a debugger to get the stack trace, it is an extra step and I'd prefer to just have the log immediately if possible. Of course, if it does become a problem, we can remove it.

File: crates/fuzzing/src/oom.rs (hunk 2 above)
Comment (fitzgen) [reply crates/fuzzing/src/oom.rs:102]: Agreed that this is pretty strict, but I think we should at least aim high and see if it is really too onerous before backing down. Especially as the effort begins, we are going to be testing smaller pieces of code where it probably is very reasonable to match this strict behavior. As we start testing larger and larger amounts of code at the same time, then we can loosen things up a bit, and when we do that, I'd like to do it via adding config knobs to `OomTest` rather than loosening the requirements for all existing tests.

But yeah I'd imagined that when we are in the "non-strict" mode, we would treat `DidOom` as `AlwaysOomFromNowOn`, if we wanted to allow allocation to continue. If we want to allow subsequent allocations to potentially succeed instead, then I think we would want to enforce an available bytes limit, and let deallocations raise the available bytes again. Which is an interesting idea. Perhaps worth investigating in the future (but not in this PR obviously).

File: crates/fuzzing/src/oom.rs (hunk 3 above)
Comment (fitzgen) [reply crates/fuzzing/src/oom.rs:215]: > This feels perhaps a bit over-ambitious to me where we might have various edge cases where allocation failure can be handled.

Again, I'd prefer to start ambitious, and have that be the default, and then add knobs to loosen things up and be less strict that tests opt into.

> For example we might have a test saying "invoke a function twice" which invokes it a second time even if the first time failed, ensuring that an OOM on one call doesn't prevent some other call from happening (and reading corrupted state/etc). In such a situation the test would still return success, despite handling the OOM.

Right and this kind of test is one where we would definitely twist that knob.

But, at least how I am imagining things now, most (all?) of these "retry" code paths will be in embedder code, not in the core Wasmtime APIs themselves. For example, instance allocation will just return an error on OOM, not try again N times or attempt to free anything before trying again; all of that logic really belongs in the embedder, IMO. So probably most of our tests can actually work in the stricter mode (but we should also test retrying to exercise things for when the embedder will do it).

File: crates/fuzzing/src/oom.rs (hunk 4 above)
Comment (fitzgen) [reply crates/fuzzing/src/oom.rs]: That's a great point, will do in a minute.
